pkg/playback/interactions.go

```diff
@@ -0,0 +1,183 @@
+	writer         io.Writer
+	interactions   []cassettePair
+	reqSerializer  serializer
+	respSerializer serializer
+}
+
+func newInteractionRecorder(writer io.Writer, reqSerializer serializer, respSerializer serializer) (recorder *interactionRecorder, err error) {
+	recorder = &interactionRecorder{}
+
+	recorder.writer = writer
+	recorder.reqSerializer = reqSerializer
+	recorder.respSerializer = respSerializer
+
+	recorder.interactions = make([]cassettePair, 0)
+
+	return recorder, nil
+}
+
+// takes a generic struct
+func (recorder *interactionRecorder) write(typeOfInteraction interactionType, req interface{}, resp interface{}) error {
+	reqBytes, err := recorder.reqSerializer(req)
+
+	if err != nil {
+		return err
+	}
+
+	respBytes, err := recorder.respSerializer(resp)
+
+	if err != nil {
+		return err
+	}
+
+	recorder.interactions = append(recorder.interactions, cassettePair{Type: typeOfInteraction, Request: reqBytes, Response: respBytes})
+
+	return err
+}
+
+func (recorder *interactionRecorder) saveAndClose() error {
+	cassette := cassetteYaml{}
+	cassette.Interactions = recorder.interactions
+
+	yamlBytes, err := yaml.Marshal(cassette)
+	if err != nil {
+		return err
+	}
+
+	_, err = recorder.writer.Write(yamlBytes)
+	return err
+}
+
+// A interactionReplayer contains a set of recorded interactions and exposes
+// functionality to play them back.
+type interactionReplayer struct {
+	cursor           int
+	comparator       requestComparator
+	cassette         cassetteYaml
+	respDeserializer deserializer
+	reqDeserializer  deserializer
+}
+
+func newInteractionReplayer(reader io.Reader, reqDeserializer deserializer, respDeserializer deserializer, comparator requestComparator) (replayer *interactionReplayer, err error) {
+	replayer = &interactionReplayer{}
+	replayer.cursor = 0
+	replayer.comparator = comparator
+	replayer.reqDeserializer = reqDeserializer
+	replayer.respDeserializer = respDeserializer
+
+	yamlBuf, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return replayer, err
+	}
+
+	err = yaml.Unmarshal(yamlBuf, &replayer.cassette)
+	if err != nil {
+		return replayer, err
+	}
+	return replayer, nil
+}
+
+func (replayer *interactionReplayer) write(req interface{}) (resp *interface{}, err error) {
+	if len(replayer.cassette.Interactions) == 0 {
+		return nil, errors.New("nothing left in cassette to replay")
+	}
+
+	var lastAccepted interface{}
+	acceptedIdx := -1
+
+	for idx, val := range replayer.cassette.Interactions {
+		var reader io.Reader = bytes.NewReader(val.Request)
+		requestStruct, err := replayer.reqDeserializer(&reader)
+		if err != nil {
+			return nil, fmt.Errorf("error when deserializing cassette: %w", err)
+		}
+
+		accept, shortCircuit := replayer.comparator(requestStruct, req)
+
+		if accept {
+			var reader io.Reader = bytes.NewReader(val.Response)
+			responseStruct, err := replayer.respDeserializer(&reader)
+
+			if err != nil {
+				return nil, errors.New("error when deserializing cassette")
```

Comment:
this error could be a bit better if it matched line 134

Reply:
done!